### PR TITLE
Combobox: Add a warning for combobox when exceeding the recommended amount of options

### PR DESCRIPTION
**Pull request by bohandley**
**What is this?**

This is to log a warning when a component exceeds the recommended amount of options. Ideally, we can use Faro to communicate with the team that uses a component somewhere in a situation where the component receives too many options.

**Why?**

Historically, Grafana has just crashed when filling select components that may have hundreds of thousands of options. The combobox is amazing, but it should emit some warning that can be captured or noticed (Faro?). In a perfect world, we could log a warning when the recommended amount of options is exceeded _and_ alert the team that used that component. The team could then act before an issue is created or escalation due to slowness or the browser runs out of memory.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Comment by tskarhed** [comment]
@bohandley I think this is a good combination with graceful degradation. At the moment setting the limit is a bit arbitrary, as it does well with 1M+ items. Maybe we can start with logging a warning, then investigate if/when we actually need to degrade functionality (leave options out above a certain threshold)?

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 1]
```diff
@@ -259,6 +259,11 @@ export const Combobox = <T extends string | number>({
       ? 'search'
       : 'angle-down';
 
+  if (items.length > 100) {
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:262]:
```suggestion
  if (items.length > 100_000) {
```
I think we can bump it to this at the very least. If loading this many options into Combobox is an issue, it is not related to the component itself. Also, we might want to extract this into a const `RECOMMENDED_ITEMS_AMOUNT`

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 2]
```diff
@@ -259,6 +259,11 @@ export const Combobox = <T extends string | number>({
...
+    const msg = `There are ${items.length} loaded into the component. This exceeds the recommended amount.`;
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:263]:
```suggestion
    const msg = `[Combobox] Items exceeds the recommended amount ${RECOMMENDED_ITEMS_AMOUNT}.`;
```
We don't want Faro to report separate error messages for all different values. I'm not sure if it manages to pick up the pattern.

**Comment by joshhunt** [comment]
@bohandley Nice - I've been thinking about adding exactly this, so thanks for it.

What do you think about adding this in dev builds only (`process.env.NODE_ENV !== 'production'`)? Do you think it would still be caught at useful times? Just trying to think about ensuring there's the right amount of noise at the right time 👍 

I do think it shouldn't be logged on every render though. Maybe something like:

```ts
const warningLogRef = useRef<boolean>(false);
if (!warningLogRef.current && items > 100_000) {
  console.warn('This is a warning message');
  warningLogRef.current = true;
}
```

Alternatively, it could be in the `setItems()` wrapper that'll be there soon https://github.com/grafana/grafana/pull/96054/files#diff-2fb3d852c952f79b0cb6ce6b9915205cb365d99dbabd728b898388f1714ae9eaR124.

That way it's only called when the list changes?

**Comment by bohandley** [comment]
> What do you think about adding this in dev builds only (`process.env.NODE_ENV !== 'production'`)? Do you think it would still be caught at useful times? Just trying to think about ensuring there's the right amount of noise at the right time 👍

It makes sense to have less noise, but these issues have more often than not been caught in production. We get a new customer and everything looks good and then a month later we get an escalation saying the customer has been experiencing slow load times for a query editor. I would like this to be more of an alert that teams can monitor, like whoever is on call gets a ping that their frontend is getting close to slowing down or breaking and they need to throttle, limit, use async, etc., for the options before an escalation comes in. 🦸🏻

**Comment by bohandley** [comment]
👀 Keeping an eye on https://github.com/grafana/grafana/pull/96054

I think doing this in the setItems wrapper sounds good!

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 3]
```diff
@@ -75,6 +79,26 @@ function itemFilter<T extends string | number>(inputValue: string) {
 const noop = () => {};
 const asyncNoop = () => Promise.resolve([]);
 
+function logOptions(
+  items: Array<ComboboxOption<string | number>>,
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:83]: I think just passing the count here should be enough.  You might want to put this in a separate file too, as we may use this for `MultiCombobox`.

Comment (bohandley) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:83]: Nice! I was thinking of something like this, using it in other components. If we use this for MultiCombobox, would it be possible to consider using it in the `Select` component as well?

Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:83]: I think that should be fine, as long as we make sure we don't end up drowning in logs 😅

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 4]
```diff
@@ -75,6 +79,26 @@ function itemFilter<T extends string | number>(inputValue: string) {
 const noop = () => {};
 const asyncNoop = () => Promise.resolve([]);
 
+function logOptions(
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:82]: this function should be removed in favor of the new from from utils yeah?

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx (hunk 3 above)
Comment (bohandley) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx]: I agree, we should make sure we are not logging excessively. With the RECOMMENDED_ITEMS_AMOUNT though, we should hopefully only be getting the log if someone is using a component to display an abnormally large amount of options. For example, Prometheus 🥲, which has some customers and consumers who push the boundaries for options. Once we have this in production, I will be able to look for the logs in Faro. Maybe we won't even need the warning if Faro is alerting teams that consume excessive amounts of options?

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx (hunk 4 above)
Comment (bohandley) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx]: Yes! I ran into some betterer issues and wasn't sure if I had added them so I only pushed part of the commit, just the new files. It is removed!

**Comment by bohandley** [comment]
@joshhunt with Faro and logInfo removed, we now only do the `console.warn` with extra details. I have created the issue to look into Faro capturing the warning to alert teams and to use the logging in other components. Let me know if this looks good to you all! https://github.com/grafana/grafana/issues/96495 Thank you!!!

**Review by joshhunt — APPROVED** [review]

**Pull request merged**